server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardWriteLoadDistributionMetrics.java

```diff
@@ -0,0 +1,271 @@
+                    );
+                }
+
+                final double prioritisationThreshold = BalancedShardsAllocator.Balancer.PrioritiseByShardWriteLoadComparator.THRESHOLD_RATIO
+                    * maxShardWriteLoad;
+                writeLoadPrioritisationThresholdMetrics.add(new DoubleWithAttributes(prioritisationThreshold, nodeAttrs));
+
+                final long shardsExceedingThreshold = (long) shardWeightHistogram.getCountBetweenValues(
+                    prioritisationThreshold,
+                    Double.MAX_VALUE
+                );
+                shardCountsExceedingPrioritisationThresholdMetrics.add(new LongWithAttributes(shardsExceedingThreshold, nodeAttrs));
+            }
+            shardWriteLoadSumMetrics.add(new DoubleWithAttributes(totalShardWriteLoad, nodeAttrs));
+        }
+
+        for (int i = 0; i < trackedPercentiles.length; i++) {
+            lastWriteLoadDistributionMetrics.set(i, writeLoadDistributionMetrics[i]);
+        }
+        lastWriteLoadPrioritisationThresholdMetrics.set(writeLoadPrioritisationThresholdMetrics);
+        lastShardCountExceedingPrioritisationThresholdMetrics.set(shardCountsExceedingPrioritisationThresholdMetrics);
+        lastWriteLoadSumMetrics.set(shardWriteLoadSumMetrics);
+        lastMetricsCollected = false;
+    }
+
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    private List<DoubleWithAttributes>[] createPercentileArrays(int percentileCount, int ingestNodeCount) {
+        List<DoubleWithAttributes>[] lists = new List[percentileCount];
+        for (int i = 0; i < percentileCount; i++) {
+            lists[i] = new ArrayList<>(ingestNodeCount);
+        }
+        return lists;
+    }
+
+    /**
+     * Get the metric name for the shard write load distribution metric for the
+     * specified percentile.
+     *
+     * @param percentile The desired percentile
+     * @return The metric name
+     */
+    public static String shardWriteLoadDistributionMetricName(int percentile) {
+        return Strings.format(WRITE_LOAD_DISTRIBUTION_METRIC_NAME, percentile);
+    }
+
+    /**
+     * We only calculate metrics for indexing nodes.
+     * <p>
+     * Note this means it only works for stateless currently
+     */
+    private boolean isIndexingNode(DiscoveryNode discoveryNode) {
+        return discoveryNode.getRoles().contains(DiscoveryNodeRole.INDEX_ROLE);
+    }
+
+    private double roundTinyValuesToZero(double value) {
+        assert value >= 0.0 : "We got a negative write load?! " + value;
+        return value < 0.01 ? 0.0 : value;
+    }
+
+    private Map<String, Object> getAttributesForNode(DiscoveryNode node) {
+        return Map.of("es_node_id", node.getId(), "es_node_name", node.getName());
+    }
+
+    // visible for testing
+    final Collection<DoubleWithAttributes> getWriteLoadDistributionMetrics(int index) {
+        lastMetricsCollected = true;
```

Comment:
This flag needs to be set after fetching and clearing the data, in all the methods,

```suggestion
        var metrics = lastWriteLoadDistributionMetrics.getAndSet(index, List.of());
        lastMetricsCollected = true;
        return metrics;
```
otherwise APM sets the flag, onNewInfo unsets the flag, then APM actually collects the data, leaving the flag unset and nothing to collect (won't get refilled until APM calls again to trip the flag, and get nothing)

Reply:
Yeah you're right its backwards. Fixed in 12c6b566e12704ef3a4b05bb042283ac9ac29f30